File: core/src/androidTest/java/com/emarsys/core/request/model/CompositeRequestModelTest.java
package com.emarsys.core.request.model;

import com.emarsys.core.testUtil.TimeoutUtils;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;

import java.util.HashMap;

import static junit.framework.Assert.assertEquals;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.not;

public class CompositeRequestModelTest {

    public static final int TIMESTAMP = 800;
    public static final int TTL = 1000;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Test
    public void testEquals_withEqualModels() {
        CompositeRequestModel model1 = new CompositeRequestModel(
                "https://google.com",
                RequestMethod.GET,
                new HashMap<String, Object>() {{
                    put("payload_key1", 6);
                    put("payload_key2", false);
                    put("payload_key3", "value");
                }},
                new HashMap<String, String>() {{
                    put("header_key1", "value1");
                    put("header_key2", "value2");
                    put("header_key3", "value3");
                }},
                TIMESTAMP,
                TTL,
                new String[]{"child_id1", "child_id2", "child_id3", "child_id4"});

        CompositeRequestModel model2 = new CompositeRequestModel(
                "https://google.com",
                RequestMethod.GET,
                new HashMap<String, Object>() {{
                    put("payload_key1", 6);
                    put("payload_key2", false);
                    put("payload_key3", "value");
                }},
                new HashMap<String, String>() {{
                    put("header_key1", "value1");
                    put("header_key2", "value2");
                    put("header_key3", "value3");
                }},
                TIMESTAMP,
                TTL,
                new String[]{"child_id1", "child_id2", "child_id3", "child_id4"});

        assertEquals(model1, model2);
    }

    @Test
    public void testEquals_withDifferentChildIds() {
        String url = "https://google.com";
        RequestMethod method = RequestMethod.GET;
        HashMap<String, Object> payload = new HashMap<String, Object>() {{
            put("payload_key1", 6);
            put("payload_key2", false);
            put("payload_key3", "value");
        }};
        HashMap<String, String> headers = new HashMap<String, String>() {{
            put("header_key1", "value1");
            put("header_key2", "value2");
            put("header_key3", "value3");
        }};

        CompositeRequestModel model1 = new CompositeRequestModel(
                url,
                method,
                payload,
                headers,
                TIMESTAMP,
                TTL,
                new String[]{"child_id4"});

        CompositeRequestModel model2 = new CompositeRequestModel(
                url,
                method,
                payload,
                headers,
                TIMESTAMP,
                TTL,
                new String[]{"child_id1", "child_id2", "child_id3"});

        assertThat(model1, not(equalTo(model2)));
    }

}

File: core/src/androidTest/java/com/emarsys/core/request/model/CompositeRequestModelTest.kt
package com.emarsys.core.request.model

import com.emarsys.core.provider.timestamp.TimestampProvider
import com.emarsys.core.provider.uuid.UUIDProvider
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.net.URL

class CompositeRequestModelTest {

    private companion object {
        const val TIMESTAMP = 800L
        const val TTL = 1000L
        const val URL = "https://emarsys.com"
        const val ID = "id"
        val METHOD = RequestMethod.PUT
        val ORIGINAL_IDS = arrayOf("uuid")
    }



    private lateinit var timestampProvider: TimestampProvider
    private lateinit var uuidProvider: UUIDProvider
    private lateinit var payload: Map<String, Any>
    private lateinit var headers: Map<String, String>

    @BeforeEach
    fun setUp() {
        payload = createPayload()
        headers = createHeaders()
        timestampProvider = TimestampProvider()
        uuidProvider = UUIDProvider()
    }

    @Test
    fun testEquals_withEqualModels() {
        val model1 = CompositeRequestModel(
                ID,
                "https://google.com",
                RequestMethod.GET,
                mapOf("payload_key1" to 6,
                        "payload_key2" to false,
                        "payload_key3" to "value"
                ),
                mapOf("header_key1" to "value1",
                        "header_key2" to "value2",
                        "header_key3" to "value3"
                ),
                TIMESTAMP,
                TTL,
                arrayOf("child_id1", "child_id2", "child_id3", "child_id4"))

        val model2 = CompositeRequestModel(
                ID,
                "https://google.com",
                RequestMethod.GET,
                mapOf(
                        "payload_key1" to 6,
                        "payload_key2" to false,
                        "payload_key3" to "value"
                ),
                mapOf(
                        "header_key1" to "value1",
                        "header_key2" to "value2",
                        "header_key3" to "value3"
                ),
                TIMESTAMP,
                TTL,
                arrayOf("child_id1", "child_id2", "child_id3", "child_id4"))

        model1 shouldBe model2
    }

    @Test
    fun testEquals_withDifferentChildIds() {
        val url = "https://google.com"
        val method = RequestMethod.GET
        val payload = mapOf(
                "payload_key1" to 6,
                "payload_key2" to false,
                "payload_key3" to "value"
        )
        val headers = mapOf(
                "header_key1" to "value1",
                "header_key2" to "value2",
                "header_key3" to "value3"
        )
        val model1 = CompositeRequestModel(
                ID,
                url,
                method,
                payload,
                headers,
                TIMESTAMP,
                TTL,
                arrayOf("child_id4"))

        val model2 = CompositeRequestModel(
                ID,
                url,
                method,
                payload,
                headers,
                TIMESTAMP,
                TTL,
                arrayOf("child_id1", "child_id2", "child_id3"))
        model1 shouldNotBe model2
    }

    @Test
    fun testBuilder_mandatoryArgumentsInitialized() {
        val result = CompositeRequestModel.Builder(timestampProvider, uuidProvider)
                .url(URL)
                .build()

        result.url shouldBe URL(URL)
    }

    @Test
    fun testBuilder_originalIdsCorrectlySet() {
        val result = CompositeRequestModel.Builder(timestampProvider, uuidProvider)
                .url(URL)
                .originalRequestIds(arrayOf("1"))
                .build()

        result.originalRequestIds shouldBe arrayOf("1")
    }

    @Test
    fun testBuilder_from() {
        val expected = CompositeRequestModel(ID, URL, METHOD, payload, headers, TIMESTAMP, TTL, ORIGINAL_IDS)

        val result = CompositeRequestModel.Builder(expected).build()

        result shouldBe expected
    }

    private fun createPayload(): Map<String, Any> {
        return mutableMapOf("key1" to "value1", "key2" to mapOf("key3" to "value3", "key4" to 5))
    }

    private fun createHeaders(): Map<String, String> {
        val result = HashMap<String, String>()
        result["content"] = "application/x-www-form-urlencoded"
        result["accept"] = "application/json"
        return result
    }

}